### PR TITLE
[amp-story-player] 🐛 fix signalReady

### LINKED ISSUE #31445: [amp-story-player] Certain stories fail to 'show()' at load time
## What's the issue?

I am creating a player with `n` stories. Under some cases I do not want the first story to be played first, but some other story in the list. say, the third story. 

When the player is **ready**, I call  `show()` on the story I want to play. For some stories it works for others it does not. 

e.g. https://jsfiddle.net/qdn9h7L8/54/

In this fiddle, I have the following stories
```html
<amp-story-player style="width: 360px; height: 600px;">
      <a href="https://preview.amp.dev/documentation/examples/introduction/stories_in_amp/">Stories in AMP - Hello World</a>
      <a href="https://web-player-cards-carousel.web.app/examples/amp-story/AMPFestPlayerDemo/s4/index.html"></a>
      <a href="https://stories-staging.slickstream.com/preet1/d/webstory/legra">Legra</a>
      <a href="https://web-player-cards-carousel.web.app/examples/amp-story/AMPFestPlayerDemo/s3/index.html"></a>
</amp-story-player>
```

If I call `show()` on the 4th story on ready, it all works
If I call `show()` on the 2nd story on ready, it does not work with a render timeout error
If I call `show()` on the 3rd story on ready, it does not work with an error:
```
reason: TypeError: Cannot read property 'then' of undefined at Yb (https://cdn.ampproject.org/amp-story-player-v0.js:69:36) at https://cdn.ampproject.org/amp-story-player-v0.js:76:163
message: "Cannot read property 'then' of undefined"
stack: "TypeError: Cannot read property 'then' of undefined↵    at Yb (https://cdn.ampproject.org/am
```

## How do we reproduce the issue?

Go to https://jsfiddle.net/qdn9h7L8/54/

In the code, update the `storyToStart` to one of the urls that do not work. (Commented out code)

## What browsers are affected?

All

## Which AMP version is affected?

Powered by AMP ⚡ HTML – Version 2011200012001

### PR DESCRIPTION
**Pull request by Enriqe**
Details / fixes #31445

When using the player APIs, calling `player.load()` and `player.show()` will cause errors (`reason: TypeError: Cannot read property 'then' of undefined
    at AmpStoryPlayer.waitForStoryToLoadPromise.`) 

This is because we are incorrectly signaling that the player is ready for interaction when it isn't in reality. The signal event and property are dispatched/set BEFORE the `build and layoutCallback()` finish. This causes race conditions when calling e.g. `show()` right after `player.load()` currently:

E.g.

```javascript
player.load();

player.addEventListener("ready", () => {
  show('url'); // this is running before build/layoutCallback() finishes.
});

if (player.isReady) {
  show('url'); // this is running before build/layoutCallback() finishes.
}
```

The proposed change is to wait at least until the first story is prerendered (once everything has been built and setup) before signaling that the player is ready.

### CODE REVIEW

File: src/amp-story-player/amp-story-player-impl.js [hunk 1]
```diff
@@ -724,16 +723,18 @@ export class AmpStoryPlayer {
     this.prerenderStories_();
     this.initializeVisibleIO_();
 
+    this.prerenderFirstStoryDeferred_.promise.then(() => {
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:726]: Is this the right heuristic? The fact that a story is prerendered and the fact that you can use show/add/etc is totally unrelated.
We should make the APIs available ASAP

**Comment by mszylkowski** [comment]
Let's say that you want to start a player showing the 3rd story in the player, would this prerender the 1st page even though you might never show it? (might be related to @gmajoulet's comment)

**Comment by gmajoulet** [comment]
> Let's say that you want to start a player showing the 3rd story in the player, would this prerender the 1st page even though you might never show it? (might be related to @gmajoulet's comment)

This would be a different code path, we don't need to worry about this specific example for now. We'll give proper guidance and control over the prerendering experience for people trying to do that.

**Comment by pshihn** [comment]
> We'll give proper guidance and control over the prerendering experience for people trying to do that.

This is definitely a common use case. Perhaps create an issue to provide whatever the proper guidance and control is?

**Comment by Enriqe** [comment]
Closed in favor of #32153

**Pull request closed without merging**